OrbitCore/CallstackData.cpp

```diff
@@ -32,5 +117,51 @@ void CallstackData::AddCallStackFromKnownCallstackData(
     absl::MutexLock lock(&unique_callstacks_mutex_);
     unique_callstacks_[hash] = std::move(unique_callstack);
   }
-  callstack_events_.push_back(CallstackEvent(event));
+  absl::MutexLock lock(&callstack_events_by_tid_mutex_);
+  callstack_events_by_tid_[event.thread_id()][event.time()] = CallstackEvent(event);
+}
+
+const CallStack* CallstackData::GetCallStack(CallstackID callstack_id) const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  auto it = unique_callstacks_.find(callstack_id);
+  if (it != unique_callstacks_.end()) {
+    return it->second.get();
+  }
+  return nullptr;
+}
+
+bool CallstackData::HasCallStack(CallstackID callstack_id) const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  return unique_callstacks_.count(callstack_id) > 0;
+}
+
+void CallstackData::ForEachUniqueCallstack(
+    const std::function<void(const CallStack&)>& action) const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  for (const auto& it : unique_callstacks_) {
+    action(*it.second);
+  }
+}
+
+void CallstackData::ForEachFrameInCallstack(uint64_t callstack_id,
+                                            const std::function<void(uint64_t)>& action) const {
+  absl::MutexLock lock(&unique_callstacks_mutex_);
+  for (uint64_t frame : unique_callstacks_.at(callstack_id)->GetFrames()) {
+    action(frame);
+  }
+}
+
+absl::flat_hash_map<CallstackID, std::shared_ptr<CallStack>>
+CallstackData::GetUniqueCallstacksCopy() const {
```

Comment:
Is that still called/relevant? What is the user? Seems pretty expensive.

Reply:
It's used in `OrbitApp` and the result is passed to the constructor of `SamplingProfiler`, as `SamplingProfiler` holds its own `unique_callstacks_`. We might want to change this but it's probably part of a `SamplingProfiler` rework that we should do (even the name is obsolete...), but it's way out of scope here.

Reply:
Looks like we still use it to create the sample report on capture complete. Do you think we can get rid of it?

Reply:
@dpallotti Okay, thanks for clarification. Yes should be targeted when doing SamplingProfiler refactoring.